parsel/csstranslator.py

```diff
@@ -38,46 +49,77 @@ def __str__(self):
 
         return path
 
-    def join(self, combiner, other, *args, **kwargs):
+    def join(
+        self: "Self",
+        combiner: str,
+        other: OriginalXPathExpr,
+        *args: Any,
+        **kwargs: Any,
+    ) -> "Self":
+        if not isinstance(other, XPathExpr):
+            raise ValueError(
+                f"Expressions of type {__name__}.XPathExpr can ony join expressions"
+                f" of the same type (or its descendants), got {type(other)}"
+            )
         super().join(combiner, other, *args, **kwargs)
         self.textnode = other.textnode
         self.attribute = other.attribute
         return self
 
 
+if TYPE_CHECKING:
+    # requires Python 3.8
+    from typing import Protocol
```

Comment:
Did you intent to import from `typing_extensions`, then?

I wonder if things like this would justify running mypy with minimum dependencies and Python versions.

Reply:
> Did you intent to import from typing_extensions, then?

`types-lxml` is not available for 3.7 (I didn't check if it's possible to change that) so we already don't support running mypy on 3.7.